src/plugin/nerd_plugin.js

```diff
@@ -166,122 +218,141 @@ class NerdPluginVirtual extends EventEmitter {
-        })
-      }).catch(this._handle)
-    })
+    const valid = yield this.balance.checkAndSettleOutgoing(transfer.amount)
+    const validAmount = (typeof transfer.amount === 'string' &&
+      !isNaN(transfer.amount - 0) && (transfer.amount - 0 >= 0))
+    const validAccount = (typeof transfer.account === 'string')
+
+    if (!validAmount || !validAccount) {
+      throw new InvalidFieldsError('missing amount or account')
+    }
+
+    if (!valid) {
+      throw new NotAcceptedError('amount not valid with current balance')
+    }
+
+    const settleAddress = this.settler && (yield this.settler.getAccount())
+    yield this.rpc.call('send', [transfer, settleAddress])
+
+    if (!transfer.executionCondition) {
+      yield this.balance.add(transfer.amount)
+      this.emit('outgoing_transfer', transfer)
+    } else {
+      this.emit('outgoing_prepare', transfer)
+    }
+
+    yield this._completeTransfer(transfer)
+    this._handleTimer(transfer)
+
+    return Promise.resolve(null)
   }
 
   getInfo () {
     return Promise.resolve(this.info)
   }
 
   getFulfillment (transferId) {
-    return this._getFulfillment(transferId).then((fulfillment) => {
-      if (!fulfillment) {
-        return Promise.reject(null)
-      } else {
-        return Promise.resolve(fulfillment)
-      }
-    })
+    return co.wrap(this._getFulfillment).call(this, transferId)
   }
 
-  _getFulfillment (transferId) {
-    return this.transferLog.getFulfillment(transferId)
+  * _getFulfillment (transferId) {
+    const stored = yield this.transferLog.get(transferId)
+    if (!stored) {
+      throw new TransferNotFoundError(transferId + ' not found')
+    }
+
+    const fulfillment = yield this.transferLog.getFulfillment(transferId)
+    if (!fulfillment) {
+      throw new MissingFulfillmentError(transferId + ' has not been fulfilled')
+    }
+
+    return fulfillment
   }
 
   fulfillCondition (transferId, fulfillmentBuffer) {
     let fulfillment = fulfillmentBuffer.toString()
-    let transfer = null
     this._log('fulfilling: ' + fulfillment)
-    return this.transferLog.get(transferId).then((storedTransfer) => {
-      transfer = storedTransfer
-      return this._fulfillConditionLocal(transfer, fulfillment)
-    }).catch(this._handle)
+
+    return co.wrap(this._fulfillConditionLocal).call(this, transferId, fulfillment)
+      .then((res) => {
+        this.rpc.call('fulfill', [res.transfer, res.direction])
+        return Promise.resolve(null)
+      })
   }
 
-  rejectIncomingTransfer (transferId) {
-    let transfer = null
+  * _fulfillConditionLocal (transferId, fulfillment) {
+    const transfer = yield this.transferLog.get(transferId)
+
+    if (!transfer || !transfer.executionCondition) {
+      throw new TransferNotFoundError('no conditional transfer exists with the given id')
+    }
+
+    const fulfilled = yield this.transferLog.isFulfilled(transfer.id)
+    if (fulfilled) {
+      throw new RepeatError('this transfer has already been fulfilled')
```

Comment:
Trying to fulfill an already fulfilled transfer should succeed, not fail. (Idempotence.)